x/transfer/ibc_handlers.go

```diff
@@ -138,25 +75,55 @@ func (im IBCModule) HandleTimeout(ctx sdk.Context, packet channeltypes.Packet, r
 	if err != nil {
 		return errors.Wrapf(sdkerrors.ErrInvalidAddress, "failed to decode address from bech32: %v", err)
 	}
+	if !im.ContractManagerKeeper.HasContractInfo(ctx, senderAddress) {
+		return nil
+	}
 
 	cacheCtx, writeFn, newGasMeter := im.createCachedContext(ctx)
-	defer im.outOfGasRecovery(ctx, newGasMeter, senderAddress, packet, data, "timeout")
 
-	_, err = im.ContractManagerKeeper.SudoTimeout(cacheCtx, senderAddress, packet)
+	// distribute fee
+	im.wrappedKeeper.FeeKeeper.DistributeTimeoutFee(ctx, relayer, feetypes.NewPacketID(packet.SourcePort, packet.SourceChannel, packet.Sequence))
+	func() {
+		// early error initialisation, to choose a correct `if` branch right after the closure, in case of successfully `out of gas` panic recovered
+		// if SudoTimeout successful, then `err` is set to `nil`
+		defer im.outOfGasRecovery(newGasMeter, &err)
+		_, err = im.ContractManagerKeeper.SudoTimeout(cacheCtx, senderAddress, packet)
+	}()
+
 	if err != nil {
+		// the contract either returned an error or panicked with `out of gas`
 		im.ContractManagerKeeper.AddContractFailure(ctx, packet.SourceChannel, senderAddress.String(), packet.GetSequence(), "timeout")
 		im.keeper.Logger(ctx).Debug("failed to Sudo contract on packet timeout", err)
 	} else {
-		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
 		writeFn()
 	}
 
-	// distribute fee only if the sender is a contract
-	if im.ContractManagerKeeper.HasContractInfo(ctx, senderAddress) {
-		im.wrappedKeeper.FeeKeeper.DistributeTimeoutFee(ctx, relayer, feetypes.NewPacketID(packet.SourcePort, packet.SourceChannel, packet.Sequence))
-	}
-
-	ctx.GasMeter().ConsumeGas(newGasMeter.GasConsumed(), "consume from cached context")
+	ctx.GasMeter().ConsumeGas(newGasMeter.GasConsumedToLimit(), "consume gas from cached context")
 
 	return nil
 }
+
+// outOfGasRecovery converts `out of gas` panic into an error
+// leave unprocessed any other kinds of panics
+func (im IBCModule) outOfGasRecovery(
+	gasMeter sdk.GasMeter,
+	err *error,
+) {
+	if r := recover(); r != nil {
+		_, ok := r.(sdk.ErrorOutOfGas)
+		if !ok || !gasMeter.IsOutOfGas() {
+			panic(r)
+		}
+		*err = errors.Wrapf(errors.ErrPanic, "%v", r)
+	}
+}
+
+// createCachedContext creates a cached context for handling Sudo calls to CosmWasm smart-contracts.
+// If there is an error during Sudo call, we can safely revert changes made in cached context.
+func (im IBCModule) createCachedContext(ctx sdk.Context) (sdk.Context, func(), sdk.GasMeter) {
+	cacheCtx, writeFn := ctx.CacheContext()
+	sudoLimit := im.ContractManagerKeeper.GetParams(ctx).SudoCallGasLimit
+	gasMeter := sdk.NewGasMeter(sudoLimit)
+	cacheCtx = cacheCtx.WithGasMeter(gasMeter)
+	return cacheCtx, writeFn, gasMeter
+}
```

Comment:
We have two identical methods: one in ICTX module and another one in transfer module. Is there any way to avoid copy-paste and make these methods reusable on both modules?